lib/ASTGen/Sources/ASTGen/BuiltinPound.swift

```diff
@@ -0,0 +1,218 @@
+//===--- BuiltinPound.swift -----------------------------------------------===//
+//
+// This source file is part of the Swift.org open source project
+//
+// Copyright (c) 2025 Apple Inc. and the Swift project authors
+// Licensed under Apache License v2.0 with Runtime Library Exception
+//
+// See https://swift.org/LICENSE.txt for license information
+// See https://swift.org/CONTRIBUTORS.txt for the list of Swift project authors
+//
+//===----------------------------------------------------------------------===//
+
+import ASTBridging
+import BasicBridging
+@_spi(RawSyntax) import SwiftSyntax
+
+extension ASTGenVisitor {
+  /// 3 state result:
+  ///  * successfully generated: `.generated(.some(T))`
+  ///  * handled but failed: `.generated(.none)`
+  ///  * fallback to default behavior: `.ignored`
+  enum MaybeGenerated<T> {
+    case generated(T?)
+    case ignored
+  }
+
+  func maybeGenerateBuiltinPound(macroExpansionDecl node: MacroExpansionDeclSyntax) -> MaybeGenerated<BridgedASTNode> {
+    let result = maybeGenerateBuiltinPound(freestandingMacroExpansion: node)
+
+    switch result {
+    case .generated(_):
+      guard node.attributes.isEmpty && node.modifiers.isEmpty else {
+        // TODO: Diagnose.
+        fatalError("attributes not applied")
+      }
+      break
+    case .ignored:
+      break
+    }
+
+    return result
+  }
+
+  /// Handle built in pound keywords.
+  func maybeGenerateBuiltinPound(freestandingMacroExpansion node: some FreestandingMacroExpansionSyntax) -> MaybeGenerated<BridgedASTNode> {
+    var isStmtLevel: Bool {
+      switch node.parent?.kind {
+      case .codeBlockItem, .memberBlockItem:
```

Comment:
Hmm, `memberBlockItem` isn't really "statement level", kind of feels like `.assert` should only be checking for `.codeBlockItem`, while `.error, .warning` should be checking for both. What do you think?

Reply:
Decided not to do this check here. Instead
* for `.error` and `.warning`, check it in `handlePoundDiagnostic(freestandingMacroExpansion:)`.
* for `.assert`, caller of `maybeGenerateBuiltinPound(freestandingMacroExpansion:)` should diagnose it. I.e. `generate(macroExpansionExpr:)` rejects `.stmt, .decl`